File: JavaAlgorithm/src/main/java/algo/tzashinorpu/ThirdRound/Chapter09/largestValues_515.java
package algo.tzashinorpu.ThirdRound.Chapter09;

import algo.tzashinorpu.ThirdRound.TreeNode;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.List;

public class largestValues_515 {
    public List<Integer> largestValues(TreeNode root) {
        ArrayList<Integer> res = new ArrayList<>();
        ArrayDeque<TreeNode> deque = new ArrayDeque<>();
        if (root != null) deque.offerLast(root);
        while (!deque.isEmpty()) {
            int size = deque.size();
            int max = deque.peekFirst().val;
            for (int i = 0; i < size; i++) {
                TreeNode pop = deque.pollFirst();
                max = Math.max(max, pop.val);
                if (pop.right != null) deque.offerLast(pop.right);
                if (pop.left != null) deque.offerLast(pop.left);
            }
            res.add(max);
        }
        return res;
    }
}

File: Clion/ZeroTrac/Tree/largestValues_515.cpp
//
// Created by TzashiNorpu on 10/24/2022.
//
#include "tree.h"

using namespace Tree;

vector<int> Solution::largestValues(TreeNode *root) {
  vector<int> res;
  deque<TreeNode *> q;
  if (root) q.push_back(root);
  while (!q.empty()) {
    int t = q.front()->val;
    int size = q.size();
    for (int i = 0; i < size; ++i) {
      TreeNode *front = q.front();
      t = max(t, front->val);
      q.pop_front();
      if (front->left) q.push_back(front->left);
      if (front->right)q.push_back(front->right);
    }
    res.push_back(t);
  }
  return res;
}